### ISSUE TITLE
Source code with tabs did not highlight correctly

### ISSUE THREAD

**Issue by legendiguess** [rust-lang/annotate-snippets-rs#25]
Here some example `Slice` where `t` symbol in `source` code have one space behind:
![image](https://user-images.githubusercontent.com/40376451/72680804-e5752880-3adf-11ea-80c5-aa8d0809acde.png)

That `Slice` give us that output:
![image](https://user-images.githubusercontent.com/40376451/72680825-1ead9880-3ae0-11ea-87ca-f7d7f3ed5b8e.png)

Now i change space behind `t` from space to tab(which still one symbol, so range of highlight should not be changed):
![image](https://user-images.githubusercontent.com/40376451/72680845-50befa80-3ae0-11ea-8f1f-ed30feff69b2.png)
And highlight in output is wrong now:
![image](https://user-images.githubusercontent.com/40376451/72680861-6d5b3280-3ae0-11ea-87e3-014ff1ba3358.png)

**Comment by brendanzab** [comment]
Yeah, I think annotate snippets should replace tab characters with a user configurable number of spaces. This is what I do in codespan at least!

**Comment by zbraniecki** [comment]
agree! I am swamped a bit these days but will try to debug it! PRs welcomed!

**Comment by zbraniecki** [comment]
agree! I am swamped a bit these days but will try to debug it! PRs welcomed!

**Comment by zbraniecki** [comment]
@legendiguess - can you retest against master? #27 got merged and I can't reproduce it anymore locally!

**Comment by legendiguess** [comment]
@zbraniecki 
Just retested it - problem still here:
![изображение](https://user-images.githubusercontent.com/40376451/77719431-0a69a880-7007-11ea-8fda-213d8b9101ec.png)

Here gist with minimal reproducible example https://gist.github.com/legendiguess/ccb197dac4bc77e3ae7571befc8ffbe2

**Comment by zbraniecki** [comment]
Ah, I see. Now I understand the issue.

@brendanzab 's solution is one option.

The other would be to only adjust the underline calculation to account for tabs in line, but I'm not sure if display of `"\tt"` is always 5 ascii characters.

**Comment by legendiguess** [comment]
> The other would be to only adjust the underline calculation to account for tabs in line, but I'm not sure if display of `"\tt"` is always 5 ascii characters.

`\t` move next characters to the next tabulation column. so basically it's adds from 1 to 4 ascii characters, dependent on caret position off last tabulation column

**Comment by zbraniecki** [comment]
Yeah, so it may be tricky to predict. Wondering how rustc handles that in error reporting now.

**Comment by legendiguess** [comment]
> 
> 
> Yeah, so it may be tricky to predict. Wondering how rustc handles that in error reporting now.

https://github.com/rust-lang/rust/blob/5e8897b7b51636f157630e6639b711d698e1d101/src/librustc_errors/styled_buffer.rs#L16

They don't handle `\t` in a fully "proper way"

**Comment by brendanzab** [comment]
I do this using a custom writer in `codespan-reporting`: https://github.com/brendanzab/codespan/blob/master/codespan-reporting/src/term/config.rs#L56-L84

**Comment by brendanzab** [comment]
Oh, and importantly I also measure the length of strings taking the tab width into account: https://github.com/brendanzab/codespan/blob/a77e646944765dfb338185b0ba9504500afb7304/codespan-reporting/src/term/config.rs#L35-L45

**Comment by brendanzab** [comment]
Here's the relevant tests (excuse the not-so-nice test setup!): https://github.com/brendanzab/codespan/blob/a77e646944765dfb338185b0ba9504500afb7304/codespan-reporting/tests/term.rs#L415-L542

**Comment by brendanzab** [comment]
> \t move next characters to the next tabulation column. so basically it's adds from 1 to 4 ascii characters, dependent on caret position off last tabulation column

Oh wait, I understand what you're saying now, @legendiguess - yes I don't handle this properly in `codespan-reporting` either 🤦‍♂

**Comment by tisonkun** [comment]
https://github.com/rust-lang/annotate-snippets-rs/issues/58 can be a duplicate.

**Comment by tisonkun** [comment]
> a user configurable number of spaces

When I dig into this issue yesterday, it turns out to be quite subtle even the `expand` POSIX command can produce "unexpected" result.

That said, leading `\t` (only) is a easy case, while tab in the medium can be quite complex, especially mixed with other whitespace. Generally, tab means making table according to `tabstop` so you need to translate the tab into whitespace to ensure that the whitespace width are aligned at the tabstop. But it's very unclear that even `expand` handles ` \t` and `\t ` differently within one tabstop slot.

**Comment by estebank** [comment]
For the record: in rustc we replace all `\t` unconditionally with 4 ` ` (we don't account for tabstops), and treat them as 4 chars for the purposes of underlines/span positioning, and so far it has worked better than it has any right to. No-one has complained that the tabstops are ignored, and we don't get any obvious positioning bugs.

**Comment by nicolasbarbier** [comment]
> so far it has worked better than it has any right to

I only care about things aligning on my screen, not about the exact indentation width. I never use tabs except for indentation. I guess it's the same for most people.

This solution definitely gives the best bang for the (development effort) buck.
